### ISSUE TITLE
Merging @api_params with params

### ISSUE THREAD

**Issue by joshpfosi** [Apipie/apipie-rails#400]
Hi, I am working with a fairly large Rails API (55+ models) and would not like to switch all instances of `params` => `@api_params` (an undocumented requirement to use the value processing feature). If I throw a `params!.merge @api_params` into a `before_filter`, however, `apipie_validations` may not have been called yet. Calling it `:explicitly` causing results in strange behavior.

Specifically, if I make multiple request to the same endpoint within 1 rspec example, with different paramters, `@api_params` accumulates ALL the parameters across the different requests.

```ruby
    return unless defined? apipie_validations
    apipie_validations
    params.deep_merge! @api_params
```

**Comment by iNecas** [comment]
Hi,

Might the problem be just with this part https://github.com/Apipie/apipie-rails/blob/master/lib/apipie/dsl_definition.rb#L247, where we use `@api_params ||= {}` instead of `@api_params = {}`?

To the original question, we like to avoid fiddling too much with the original `params` as might lead to unexpected behavior and untransparent logic. For example, a generic Rails developer, that is used to some behaviour of the `params` might be pretty confused why suddently something else is seeing there. Using `@api_params` should give better hints on where the data come from…

**Comment by joshpfosi** [comment]
Hmm it could be that. But that is necessary to ensure *all* params for a given endpoint are included, no? I claim that this is an actual bug in Apipie where `@api_params` needs to be reset at the beginning of `_apipie_define_validators`.

Re: background: I definitely agree with the design decision to use `@api_params` and would much prefer to use it, however, the code base I'm working in is too intricate for such a blanket code change and the feature isn't worth the time required to be more judicious about the refactor.

**Comment by iNecas** [comment]
Maybe you might want to include that logic into your project: having a before_filter on the parent controller, or making it a concern?

**Comment by joshpfosi** [comment]
Yes, I have. I guess I mischaracterized the issue. The bug I am noticing is that `@api_params` is not reset in between independent requests. This cannot be desired behavior, right?

**Comment by iNecas** [comment]
Well, it seems the case is mainly in testing, rather than actual requests,
but yes, making that independnet of the request
even in tests is just better. On the other hand, we don't see this behavior
happening in our tests.

https://github.com/Apipie/apipie-rails/blob/4b9bbd30c1233f836851b0303f55adbb93320189/spec/controllers/users_controller_spec.rb#L697-L707

Any pointers on difference between your project and our test suite? Ideally
finding a reproducer using our test-suite?

-- Ivan

**Comment by joshpfosi** [comment]
Hmm I thought it might be test specific. I don't have time to look into this more in the near future but I'll try to reproduce it later! Thanks
